src/fhirtypes/common.ts

```diff
@@ -645,6 +645,50 @@ export function getUrlFromFshDefinition(
   return `${canonical}/${fhirType}/${fshDefinition.id}`;
 }
 
+/**
+ * Determines the formal FHIR type to use to define to this entity for logical models and
+ * resources. The type for profiles and extension should not be changed. If a caret value
+ * rule has been applied to the entity's type, use the value specified in that rule.
+ * Otherwise, use the appropriate default based on the fshDefinition.
+ *
+ * @param fshDefinition - The FSH definition (Logical or Resource) that the returned type refers to
+ * @param structDef - The parent StructureDefinition for the fshDefinition
+ * @returns The type to specify in the StructureDefinition for this fshDefinition
+ */
+export function getTypeFromFshDefinition(
+  fshDefinition: Profile | Extension | Logical | Resource,
+  structDef: StructureDefinition
+): string {
+  if (fshDefinition instanceof Profile || fshDefinition instanceof Extension) {
+    return structDef.type;
+  }
+
+  for (const rule of fshDefinition.rules) {
+    if (rule instanceof CaretValueRule && rule.path === '' && rule.caretPath === 'type') {
+      // this value should only be a string, but that might change at some point
+      return rule.value.toString();
+    }
+  }
+  // Default type for logical model to the StructureDefinition url;
+  // otherwise default to the id meta property.
+  // Ref: https://chat.fhir.org/#narrow/pm-with/191469,210024,211704,239822-group/near/240237602
```

Comment:
I know there has been a lot of discussion about this that I was out of office for much of, but it seems like setting `type` to an absolute URL causes problems for the IG Publisher:
![image](https://user-images.githubusercontent.com/52747882/119844565-8c3cb680-bed6-11eb-9957-7ed85dd0686c.png)

Since the Publisher seems to expect that an element path will start with the `type`, but the element path starts with the `id` of the element, `MyLogicalPatient` in this case. Is this an issue we are aware of?

Reply:
Then of course, if you do try to put the full url in the path, the IG Publisher is mad at the illegal characters in that URL.

Reply:
Ugh.  Grahame has said on multiple occasions that it should be a URL.

<img width="764" alt="image" src="https://user-images.githubusercontent.com/2278253/119847805-49c8a900-bed9-11eb-8c1a-4f230db434d3.png">

([source](https://chat.fhir.org/#narrow/stream/179177-conformance/topic/StructureDefinition.2Etype.20for.20logical.20model.3F))


<img width="761" alt="image" src="https://user-images.githubusercontent.com/2278253/119847308-de7ed700-bed8-11eb-9d6e-b0f85821d743.png">

([source](https://chat.fhir.org/#narrow/stream/179166-implementers/topic/Logical.20models.20-.20Type.20Element))


<img width="744" alt="image" src="https://user-images.githubusercontent.com/2278253/119847546-0e2ddf00-bed9-11eb-8c00-508ae2b53d68.png">

([source](https://chat.fhir.org/#narrow/stream/179177-conformance/topic/StructureDefinition.2Etype.20for.20Logical.20Models.2FCustom.20Resources))


And the spec also says it is supported and how it relates to the paths:
<img width="1104" alt="image" src="https://user-images.githubusercontent.com/2278253/119848215-a330d800-bed9-11eb-8126-62602e03dfaf.png">

([source](http://hl7.org/fhir/R4/structuredefinition-definitions.html#StructureDefinition.type))

I'll ask on Zulip.

Reply:
@ngfreiter -- what version of the IG Publisher are you using?  I just tried this with a logical model and it worked fine.

Reply:
I was using an older version, as you've seen (and I've also confirmed) looks like this works in more recent versions. Sorry I didn't realize mine was out of date!